lib/pos_mempool.go

```diff
@@ -0,0 +1,391 @@
+	dmp.ledger = NewBalanceLedger()
+
+	// Create the persister
+	dmp.persister = NewMempoolPersister(dmp.db, 30000)
+
+	// Start the persister and retrieve transactions from the database.
+	dmp.persister.Start()
+	txns, err := dmp.persister.GetPersistedTransactions()
+	if err != nil {
+		return errors.Wrapf(err, "PosMempool.Start: Problem retrieving transactions from persister")
+	}
+	// We set the emitEvents flag to false so that we don't emit events when adding transactions from the persister.
+	dmp.emitEvents = false
+	for _, txn := range txns {
+		if err := dmp.addTransactionNoLock(txn); err != nil {
+			glog.Errorf("PosMempool.Start: Problem adding transaction with hash (%v) from persister: %v",
+				txn.Hash, err)
+		}
+	}
+	dmp.emitEvents = true
+	dmp.status = PosMempoolStatusRunning
+
+	return nil
+}
+
+func (dmp *PosMempool) Stop() {
+	if !dmp.IsRunning() {
+		return
+	}
+
+	// Close the persister and stop the database.
+	if err := dmp.persister.Stop(); err != nil {
+		glog.Errorf("PosMempool.Stop: Problem stopping persister: %v", err)
+	}
+	if err := dmp.db.Close(); err != nil {
+		glog.Errorf("PosMempool.Stop: Problem closing database: %v", err)
+	}
+	// Reset the transaction register and the ledger.
+	dmp.txnRegister.Reset()
+	dmp.ledger.Reset()
+
+	dmp.emitEvents = false
+	dmp.status = PosMempoolStatusNotRunning
+}
+
+func (dmp *PosMempool) IsRunning() bool {
+	return dmp.status == PosMempoolStatusRunning
+}
+
+// ProcessMsgDeSoTxn validates a MsgDeSoTxn transaction and adds it to the mempool if it is valid.
+// If the mempool overflows as a result of adding the transaction, the mempool is pruned.
+func (dmp *PosMempool) ProcessMsgDeSoTxn(txn *MsgDeSoTxn) error {
+	if !dmp.IsRunning() {
+		return errors.Wrapf(MempoolErrorNotRunning, "PosMempool.ProcessMsgDeSoTxn: ")
+	}
+
+	// First, validate that the transaction is properly formatted according to BalanceModel.
+	latestBlockHeight := dmp.latestBlockNode.Height
+	if err := ValidateDeSoTxnSanityBalanceModel(txn, uint64(latestBlockHeight), dmp.params, dmp.globalParams); err != nil {
+		return errors.Wrapf(err, "PosMempool.AddTransaction: Problem validating transaction sanity")
+	}
+
+	// Construct the MempoolTx from the MsgDeSoTxn.
+	mempoolTx, err := NewMempoolTx(txn, uint64(latestBlockHeight))
+	if err != nil {
+		return errors.Wrapf(err, "PosMempool.AddTransaction: Problem constructing MempoolTx")
+	}
+
+	// Check transaction signature
+	if _, err := dmp.latestBlockView.VerifySignature(txn, latestBlockHeight); err != nil {
+		return errors.Wrapf(err, "PosMempool.AddTransaction: Signature validation failed")
+	}
+
+	// Add the transaction to the mempool.
+	if err := dmp.AddTransaction(mempoolTx); err != nil {
+		return errors.Wrapf(err, "PosMempool.AddTransaction: Problem adding transaction to mempool")
+	}
+	return nil
+}
+
+// AddTransaction is the main function for adding a new transaction to the mempool.
+func (dmp *PosMempool) AddTransaction(txn *MempoolTx) error {
```

Comment:
Is there ever a scenario where an external caller will want to use AddTransaction and not ProcessMsgDeSoTxn? IMO, it shouldn't be possible for callers to add txns to the mempool, bypassing internal validations. If you agree, then we may as well rename ProcessMsgDeSoTxn to `AddTransaction(txn *MsgDeSoTxn)`, and collapse the two functions into one to simplify the interface